File: cpp/33.cpp
#include <iostream>
#include <vector>
#include <algorithm>
#include <iterator>
#include <utility>
#include <map>
#include <set>
#include <string>
#include <unordered_map>
#include <climits>
#include <numeric>
using namespace std;

template<typename T>
void prints(vector<T> &v){
  cout << "[";
  for(uint i=0; i<v.size()-1; ++i)
    cout << v[i] << ", ";
  if(v.size()) cout << v[v.size()-1];
  cout << "]\n";
}

class Solution {
public:
  int search(int A[], int n, int target) {
    // find the fist one less than A[0]
    int l = 0, r = n;
    if(A[0] > A[n-1]){
      int pivot;
      while(l+1 != r){
        int m = (l + r) / 2;
        // A[l] >= A[0] && A[r] < A[0]
        if(A[m] >= A[0])
          l = m;
        else
          r = m;
      }
      pivot = r;
      // cout << pivot << endl;
      if(target >= A[0])
        l = 0, r = pivot;
      else
        l = pivot, r = n;
    }
    if(target < A[l] || target > A[r-1]) return -1;
    while(l+1 != r){
      int m = (l + r) / 2;
      if(A[m] <= target)
        l = m;
      else
        r = m;
    }
    return A[l] == target ? l : -1;
  }
};

int main(){
  auto sol = Solution();
  // int a[] = {4,5,6,7,0,1,2};
  // int a[] = {0,1,2,4,5,6,7};
  int a[] = {1, 3};
  cout << sol.search(a, 2, 2);
}

File: python/33.py
class Solution(object):
    @staticmethod
    def bs(nums, i, j, target):
        if target < nums[i]:
            return -1
        # nums[i] <= target < nums[j]
        while i+1 < j:
            m = (i + j) / 2
            if nums[m] <= target:
                i = m
            else:
                j = m
        if i < j and nums[i] == target:
            return i
        else:
            return -1

    def search(self, nums, target):
        """
        :type nums: List[int]
        :type target: int
        :rtype: int
        """
        n = len(nums)
        i, j = 0, n
        # nums[i] >= nums[0] && nums[j] < nums[0]
        while i+1 < j:
            m = (i+j) / 2
            if nums[m] >= nums[0]:
                i = m
            else:
                j = m
        if j == n or target >= nums[0]:
            return self.bs(nums, 0, j, target)
        else:
            return self.bs(nums, j, n, target)
